src/js/background/context-menu.js

```diff
@@ -0,0 +1,388 @@
+      const matches = url.match(/^https?\:\/\/([^\/?#]+)(?:[\/?#]|$)/i);
+      // extract hostname (will be null if no match is found)
+      return matches && matches[1];
+    },  
+    getHostnameFromUrlConstructor: (url) => {
+      const { hostname } = new URL(url);
+      return hostname;
+    },  
+    getMostRecentAliases: (array)=> {
+      // Flipped to match the same order as the dashboard
+      array.reverse();
+
+      // Limit to 5
+      if (array.length > 5) { array.length = 5 };
+      return array;
+    },
+    getSiteSpecificAliases: (array, domain)=> {
+
+      // Flipped to match the same order as the dashboard
+      array.reverse();
+
+      const filteredAliases = array.filter(alias => alias.generated_for === domain);
+
+      // If 5 results for specific domain
+      if (filteredAliases.length > 5) { 
+        filteredAliases.length = 5;
+        return filteredAliases;
+      };
+
+      return filteredAliases
+    },
+    getUserStatus: {
+      generateAliases: async (numberOfAliasesCreated = null)=> {
+        const { premium } = await browser.storage.local.get("premium");
+    
+        // Short-circuit if the user is premium.
+        if (premium) return true;
+        
+        const { maxNumAliases } = await browser.storage.local.get("maxNumAliases");
+        const { relayAddresses } = await browser.storage.local.get("relayAddresses");
+    
+        let aliasesRemaining = maxNumAliases - relayAddresses.length;
+    
+        if (numberOfAliasesCreated) {
+          aliasesRemaining = maxNumAliases - numberOfAliasesCreated.length;
+        }
+        
+        // The user cannot create and additional aliases.
+        if (aliasesRemaining < 1) return false;
+    
+        return true;
+      },
+      numberOfAliases: async () => {
+        const { relayAddresses } = await browser.storage.local.get("relayAddresses");
+        return relayAddresses.length;
+      },
+      upgradeToPremium: async()=> {
+        const { premium } = await browser.storage.local.get("premium");
+      
+        // Note: If user is already premium, this will return false.
+        return !premium;
+      },
+    },
+    onCreatedCallback: ()=> {
+      // Catch errors when trying to create the same menu twice.
+      // The browser.menus API is limited. You cannot query if a menu item already exists.
+      // The error it throws does not show up to the user.
+
+      if (browser.runtime.lastError) {
+        return;
+      }
+    }
+  }
+};
+
+// TLDR: We make sure the menu is still open when performing the refresh command, as to not waste resoruces. 
+// This method is taken from https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/menus/onShown documentation. 
+let lastMenuInstanceId = 0;
+let nextMenuInstanceId = 1;
+
+// Events
+browser.menus.onShown.addListener(async (info, tab) => {
+  
+  if (!info.menuIds.includes("fx-private-relay-generate-alias") ) {
+    // No Relay menu items exist. Stop listening.
```

Comment:
If you actually want to _stop_ listening, shouldn't this need a `.removeListener`?

Reply:
`browser.menus.onShown` fires for all alt-clicks, so this filters it. This is a bit of pre-optimization, if we were to update the contexts in which Relay is shown (currently being set to the default, where it could be set to only editable fields, like inputs)